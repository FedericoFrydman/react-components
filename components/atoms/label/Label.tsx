import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import { Text, TextProps, TextWeight } from '../text/Text';
import '../../assets/styles/utilities.scss';

export type LabelSize = 'tiny' | 'small' | 'medium' | 'large';

export interface LabelProps extends Omit<TextProps, 'textSize'> {
  /**
   * Size of the text. Defaults to `medium`.
   */
  labelSize?: LabelSize;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLParagraphElement>;
}

const LabelComponent = (
  { text, labelSize = 'medium', classNames = [], ...props }: LabelProps,
  ref: LabelProps['ref'],
) => {
  const getLabelTextWeight = (): TextWeight => {
    switch (labelSize) {
      case 'tiny':
        return 'bold';
      case 'small':
        return 'normal';
      case 'medium':
      case 'large':
      default:
        return 'semibold';
    }
  };

  return (
    <Text
      {...props}
      ref={ref}
      text={text}
      classNames={classNames}
      textSize={labelSize}
      textWeight={getLabelTextWeight()}
    />
  );
};

export const Label = forwardRef<HTMLParagraphElement, LabelProps>(LabelComponent);
