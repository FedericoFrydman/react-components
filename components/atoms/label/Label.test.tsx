import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Label, LabelProps, LabelSize } from './Label';

describe('Text Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Label text="some text" classNames={['custom--class']} />);
    testId = 'test-text';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the text the provided classNames', () => {
          const text = renderLabel({ classNames });

          expect(text).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('text', () => {
      const text = 'testing text';

      it('is displayed when present', () => {
        renderLabel({ text });

        expect(screen.getByText(text)).toBeVisible();
      });
    });

    describe('labelSize', () => {
      const labelSizes: LabelSize[] = ['tiny', 'small', 'medium', 'large'];
      const text = 'Test my label size';
      let labelSizeClass: string;

      describe('when "labelSize" is not defined', () => {
        it('defaults "labelSize" to the "body"', () => {
          renderLabel({ text });

          const labelEl = screen.getByText(text);
          expect(labelEl).toHaveClass('lsux-text--medium');
        });

        function itAssignsTheRightClassForTextSize(labelSize: LabelSize) {
          describe(`when "labelSize" is "${labelSize}"`, () => {
            beforeEach(() => {
              labelSizeClass = `lsux-text--${labelSize}`;
            });

            it(`assigns the element the ${labelSizeClass} class`, () => {
              renderLabel({ labelSize, text });

              const textEl = screen.getByText(text);
              expect(textEl).toHaveClass(labelSizeClass);
            });
          });
        }

        labelSizes.forEach((labelSize) => itAssignsTheRightClassForTextSize(labelSize));
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLParagraphElement>();

    it('provides a ref to the text', () => {
      expect(renderLabel({ ref })).toEqual(ref.current);
    });
  });

  const renderLabel = (props: LabelProps = {}): HTMLElement => {
    render(<Label {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
