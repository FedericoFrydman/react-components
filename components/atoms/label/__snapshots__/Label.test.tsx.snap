// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component should render the component 1`] = `
<ForwardRef(TextComponent)
  classNames={
    Array [
      "custom--class",
    ]
  }
  text="some text"
  textSize="medium"
  textWeight="semibold"
/>
`;
